fourier/fourier.go

```diff
@@ -0,0 +1,254 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fourier
+
+// FFT implements Fast Fourier Transform and its inverse for real sequences.
+type FFT struct {
+	work []float64
+	ifac [15]int
+
+	// real temporarily store complex data as
+	// pairs of real values to allow passing to
+	// the backing code. The length of real
+	// must always be half the length of work.
+	real []float64
+}
+
+// NewFFT returns an FFT initialized for work on sequences of length n.
+func NewFFT(n int) *FFT {
+	var t FFT
+	t.Reset(n)
+	return &t
+}
+
+// Len returns the length of the acceptable input.
+func (t *FFT) Len() int { return len(t.work) / 2 }
+
+// Reset reinitializes the FFT for work on sequences of length n.
+func (t *FFT) Reset(n int) {
+	if 2*n <= cap(t.work) {
+		t.work = t.work[:2*n]
+		t.real = t.real[:n]
+	} else {
+		t.work = make([]float64, 2*n)
+		t.real = make([]float64, n)
+	}
+	rffti(n, t.work, t.ifac[:])
+}
+
+// FFT computes the Fourier coefficients of the input sequence, seq,
+// placing the result in dst and returning it. This transform is
+// unnormalized since a call to FFT followed by a call of IFFT will
+// multiply the input sequence by the length of the sequence.
+//
+// If the length of seq is not t.Len(), FFT will panic.
+// If dst is nil, a new slice is allocated and returned. If dst is not nil and
+// the length of dst does not equal t.Len()/2+1, FFT will panic.
+func (t *FFT) FFT(dst []complex128, seq []float64) []complex128 {
+	if len(seq) != t.Len() {
+		panic("fourier: sequence length mismatch")
+	}
+	if dst == nil {
+		dst = make([]complex128, t.Len()/2+1)
```

Comment:
It's weird thet t.Len() and len(seq) are both being used here for the same reason. I would pick one.

Reply:
Done